### ISSUE TITLE
bad qafiberstatus in expid 154391, tile 7231, night 20221122

### ISSUE THREAD

**Issue by sybenzvi** [desihub/desisurveyops#80]
This tile was automatically marked bad by the pipeline due to every fiber failing the QAFIBERSTATUS check, but the `desi_tile_vi` program still asked me to manually validate it, which is unusual. It could be no big deal but I'm raising the issue in case we need to dig a bit deeper.
![Screen Shot 2022-12-02 at 2 48 39 PM](https://user-images.githubusercontent.com/7385438/205374720-f8a4c622-1150-4f81-bcea-212b1cee3fa1.png)

**Comment by araichoor** [comment]
that looks weird.
digging a bit, this tile has 3 exposures, two from 20221104, and one from 20221122.
as of now, the three files `exposures-daily.csv`, `tiles-daily.csv`, and `tiles-specstatus.ecsv`, but, for whatever reason, the pipeline didn t pick up the two exposures from 20221104.

```
>>> d = Table.read("/global/cfs/cdirs/desi/survey/ops/surveyops/trunk/ops/tiles-specstatus.ecsv")
>>> d[d["TILEID"] == 7231]
<Table length=1>
TILEID SURVEY FAPRGRM FAFLAVOR  NEXP EXPTIME  TILERA TILEDEC EFFTIME_ETC EFFTIME_SPEC ... LYA_EFFTIME_DARK GOALTYPE MINTFRAC LASTNIGHT   QA     USER   OVERRIDE QANIGHT  ARCHIVEDATE
int64   str7   str16   str19   int64 float64 float64 float64   float64     float64    ...     float64        str7   float64    int64    str6   str19    int64    int64      int64   
------ ------ ------- -------- ----- ------- ------- ------- ----------- ------------ ... ---------------- -------- -------- --------- ------ -------- -------- -------- -----------
  7231   main    dark maindark     3  2720.0 135.063   4.551      1170.6       1001.1 ...            946.9     dark     0.85  20221122 unsure sybenzvi        0 20221122           0
>>> d[d["TILEID"] == 7231]["TILEID", "FAFLAVOR", "NEXP", "EXPTIME", "EFFTIME_SPEC", "LASTNIGHT"]
<Table length=1>
TILEID FAFLAVOR  NEXP EXPTIME EFFTIME_SPEC LASTNIGHT
int64   str19   int64 float64   float64      int64  
------ -------- ----- ------- ------------ ---------
  7231 maindark     3  2720.0       1001.1  20221122
>>> 
>>> d = Table.read("/global/cfs/cdirs/desi/spectro/redux/daily/tiles-daily.csv")
>>> d[d["TILEID"] == 7231]["TILEID", "FAFLAVOR", "NEXP", "EXPTIME", "EFFTIME_SPEC", "LASTNIGHT"]
<Table length=1>
TILEID FAFLAVOR  NEXP EXPTIME EFFTIME_SPEC LASTNIGHT
int64   str19   int64 float64   float64      int64  
------ -------- ----- ------- ------------ ---------
  7231 maindark     3  2720.0       1001.1  20221122
>>> d = Table.read("/global/cfs/cdirs/desi/spectro/redux/daily/exposures-daily.csv")
>>> d[d["TILEID"] == 7231]["TILEID", "FAFLAVOR", "EXPTIME", "EFFTIME_SPEC", "NIGHT", "EXPID"]
<Table length=3>
TILEID FAFLAVOR EXPTIME EFFTIME_SPEC  NIGHT   EXPID 
int64   str19   float64   float64     int64   int64 
------ -------- ------- ------------ -------- ------
  7231 maindark  1050.3        375.0 20221104 151944
  7231 maindark  1415.0        470.0 20221104 151945
  7231 maindark   254.7        156.1 20221122 154391
```

```

**Comment by araichoor** [comment]
the expid=151944 and 151945 from 20221104 are listed here: https://github.com/desihub/desisurveyops/issues/74 “Mark bad guiding EXPID as bad, and reprocess #74" (https://data.desi.lbl.gov/desi/users/schlafly/misc/badguidingtileexp.ecsv)

my 2 cents: the pipeline is rightfully ignoring those two exposures, but the `exposures-daily.csv` is not setting EFFTIME_SPEC=0 for it (which it should be, I believe), which propagates in the `tiles-daily.csv` and `tiles-specstatus.ecsv`, and creates inconsistencies downstream.

@marcelo-alvarez: makes sense?

that could explain the tiles I identified in https://desisurvey.slack.com/archives/C01HNN87Y7J/p1670109389351019, for the 20221031-20221112 nights.

**Comment by schlafly** [comment]
Thanks @araichoor , good catch!

@marcelo-alvarez , @sbailey , @akremin , a consequence of this is that there are a lot of tiles from 20221031-20221114 which the NTS thinks are done (EFFTIME_SPEC adequate) but which we are not QAing, etc., because they're not actually done because we have excluded an exposure but not gotten exposures-daily.csv updated.  We should fix this so that those tiles can be observed.

**Comment by marcelo-alvarez** [comment]
> Thanks @araichoor , good catch!
> 
> @marcelo-alvarez , @sbailey , @akremin , a consequence of this is that there are a lot of tiles from 20221031-20221114 which the NTS thinks are done (EFFTIME_SPEC adequate) but which we are not QAing, etc., because they're not actually done because we have excluded an exposure but not gotten exposures-daily.csv updated. We should fix this so that those tiles can be observed.

@schlafly I am just now catching up on this, now that daily processing is caught up. 

If I understand what's going on, it's that the NTS algorithm is erroneously including the `EFFTIME_ETC` of exposures with `LASTSTEP != 'all'` in the exposure tables when deciding whether a tile is done.  For example, `LASTSTEP='ignore'` for exposures 151944 and 151945 in the exposure table for 20221104:
```
grep -h -e 151944\,s -e 151945\,s /global/cfs/cdirs/desi/spectro/redux/daily/exposure_tables/202211/exposure_table_20221104.csv
151944,science,7231,ignore,a01236789,,,1050.3276,449.87796,main,main,dark,1000.0,dark,1.09369028885192,1.231335,0.7374236190200144,135.067292,4.54483,1,1,dark,main survey,59888.493073779,20221104,|,|,bad guiding|
151945,science,7231,ignore,a01236789,,,1414.9614,553.962952,main,main,dark,1000.0,dark,1.09369028885192,1.193183,0.6379140831481658,135.067292,4.54483,1,1,dark,main survey,59888.506047017,20221104,|,|,bad guiding|
```
and yet NTS considers the tile done. I am not familiar with how NTS works but from the discussion above it seems like it uses the tiles-daily.csv and tiles-specstatus.ecsv files, which themselves depend on `EFFTIME_SPEC` in the exposures-daily.csv file, which doesn't contain `LASTSTEP` info.  

I don't know how to include `LASTSTEP` info when exposures-daily.csv is written, but I think it could be somewhere in  `desi_night_qa`. @araichoor if you know how to fix `desi_night_qa` so that the resulting exposures-daily.csv file `EFFTIME_SPEC` is zero when `LASTSTEP != 'all'` in the exposure table, would you be willing to submit a PR on a branch that includes the fix? We could test the branch by running nightqa on a linked test prod to see if `EFFTIME_SPEC` in exposures-daily.csv looks correct and, if it does, merge into main and re-run `desi_night_qa` for all the affected nights. Please let me know if this solution won't work or if there is a better one I haven't thought of.

**Comment by schlafly** [comment]
Thanks Marcelo.  Yes, the offline pipeline needs to communicate that it has rejected an exposure to the NTS by including a row in the exposures-daily.csv file with efftime_spec = 0.  Likewise this should be reflected in the tiles-daily.csv file.  I don't think a new PR is needed---we have marked many exposures bad in this way in the past---but I don't know the mechanism for transmitting that information to those two files.

**Comment by marcelo-alvarez** [comment]
Ok, I will wait for others like @akremin @sbailey @araichoor to chime in on how bad exposure info has been transmitted to exposures-daily.csv and tiles-daily.csv in the past, since I am new to this. If doing so involves explicitly modifying those files as well in the same way I modified the exposure tables, then I think a PR so that the changes are automatically transmitted from the exposure talbe might still be helpful in the future, even if it's not necessary. Thanks for your patience on this.

**Comment by marcelo-alvarez** [comment]
@schlafly @araichoor @akremin some progress on this: I found the likely culprit, which is that I had run the tsnr afterburners that update the exposures-daily.csv files for all the reprocessed nights in parallel which seems to have resulted in different tsnr afterburner instances overwriting updates from other nights in random ways. Won't do that again!

The problem should be fixed now, since I've re-run tsnr for those nights and there is no other instance where I have run tsnr afterburners for multiple nights at the same time. The `exposures-daily.csv` and `tiles-daily.csv` now correctly indicate an efftime_spec=0  contribution from the ignored exposures, after running tsnr afterburners in serial for the 9 nights with bad guidance exposures. For example
```
>>> d = Table.read("/global/cfs/cdirs/desi/spectro/redux/daily/exposures-daily.csv")
>>> d[d["TILEID"] == 7231]["TILEID", "FAFLAVOR", "EXPTIME", "EFFTIME_SPEC", "NIGHT", "EXPID"]
<Table length=3>
TILEID FAFLAVOR EXPTIME EFFTIME_SPEC  NIGHT   EXPID
int64   str19   float64   float64     int64   int64
------ -------- ------- ------------ -------- ------
  7231 maindark  1050.3          0.0 20221104 151944
  7231 maindark  1415.0          0.0 20221104 151945
  7231 maindark   254.7        156.1 20221122 154391
>>> d = Table.read("/global/cfs/cdirs/desi/spectro/redux/daily/tiles-daily.csv")
>>> <Table length=1>
TILEID FAFLAVOR  NEXP EXPTIME EFFTIME_SPEC LASTNIGHT
int64   str19   int64 float64   float64      int64
------ -------- ----- ------- ------------ ---------
  7231 maindark     3  2720.0        156.1  20221122
```
The other entries affected by the halloween guidance issues reprocessing also correct. I am now cleaning up and re-running nightqa for those nights.

Please let me know if I can help with anything else, such as propagating changes to `tiles-specstatus.ecsv`. My apologies for all the noise and confusion on this. I have been getting up to speed in real time on unfamiliar aspects of the pipeline.

**Comment by araichoor** [comment]
great you found the explanation!
please let us know when you re done cleaning things up, and I can try to have a look then.

**Comment by schlafly** [comment]
Great, thanks Marcelo.  @araichoor , if you can verify that resolves your issue here
https://desisurvey.slack.com/archives/C01HNN87Y7J/p1670109389351019
that would be reassuring.

The most important thing is that the tiles & exposures files were updated, which means the NTS will start observing these tiles again to finish them.  They'll then get into the tiles-specstatus.ecsv when observed.

To get them into the tiles-specstatus file, we need to run desi_update_tiles_specstatus.  I think that takes an argument to cut it down only to the tiles specified in this issue; there's another mode to run it on all tiles, but last time I did that it touched way more than I wanted.

**Comment by araichoor** [comment]
mmhhh... I m not sure that it will solve this "missing folder/files" thing, which is more properly reported here: https://github.com/desihub/desispec/issues/1933.
and, from what I described there, those missing files shouldn t be in the bad guiding exposures list, ie the related exposures should still have efftime_spec>0 after this cleaning up.

thinking aloud: once the `exposures-daily.csv` and `tiles-daily.csv` files are correctly updated, maybe we would want to run some sanity check to verify that what s in the tile-qa files is consistent with those `exposures-daily.csv` and `tiles-daily.csv` files (it s the case for this example TILEID=7231, but I don t have a sense if we expect that to be true all the time); and if not, we may need to rerun some tileqa files.
does that make sense?

**Comment by schlafly** [comment]
Yes, I agree that is an important check.

**Comment by marcelo-alvarez** [comment]
@araichoor nightqa has been re-run for those, so feel free to have a look, thanks.

**Comment by marcelo-alvarez** [comment]
@schlafly Since this issue -- incorrect efftime_spec in exposures-daily.csv and tiles-daily.csv for several tiles that were reprocessed for #74 -- has been fixed, should we close it?

**Comment by schlafly** [comment]
Yes.  I haven't checked this myself, but if you or Anand are happy, please close, thanks!
